adding mipgap & shadow price to glpkJs

saeid3 opened the pull request:
This PR adds settings to solve_mip function in C language. 
I am passing the function arguments one by one and as number but a better way would be to pass them as struct. something like
```C
int solve_mip(glp_prob *P, setting_obj settings) {
	glp_iocp parm;
	glp_init_iocp(&parm);
	parm.presolve = 1;
	parm.msg_lev = msg_lev;
	parm.mip_gap = settings.mip_gap;
	parm.tm_lim = settings.tm_lim;
	return glp_intopt(P, &parm);
}
```

Also I added shadow prices to the lp solutions.

saeid3 committed:
adding mipgap & shadow price to glpkJs, recompiling the wasm files

rregue commented on src/glpk.js.c:
```diff
@@ -12,11 +12,13 @@ int solve_lp(glp_prob *P, int msg_lev) {
 }
 
-int solve_mip(glp_prob *P, int msg_lev) {
+int solve_mip(glp_prob *P, int msg_lev, float mip_gap, float tm_lim) {
```
I think the ideal behavior here would be to pass some sort of struct that that can take the exact same values as `glp_iocp` below. I am not sure how to do that though.

rregue replied:
is time limit here a float or an int?

rregue commented on src/pre.js:
```diff
@@ -98,13 +101,15 @@ var glpkPromise = new Promise(function (resolve) {
-
-					solve(P, ret.result, msg_lev);
+					
+					const settings = lp.settings
```
I don't think we should pass the settings as part of the lp object, instead we should change the function `'solve': function (lp, msg_lev) {` from line 93 to be `'solve': function (lp, options) {`

rregue commented on src/pre.js:
```diff
@@ -229,6 +235,9 @@ var glpkPromise = new Promise(function (resolve) {
 						res.vars[glp_get_col_name(P, i)] = glp_get_col_prim(P, i);
 					}
+					for (i = 1, ii = glp_get_num_rows(P); i < ii + 1; i++) {
+						res.cons[glp_get_row_name(P, i)] = {'dual': glp_get_row_dual(P, i)};
```
also, if we are adding this, we also need to update the types in https://github.com/TakeScoop/glpk.js/blob/master/glpk.d.ts

rregue replied:
I believe you do not need the `'`

saeid3 committed:
reshaping to have settings as solve function arg

rregue commented on glpk.d.ts:
```diff
@@ -17,7 +17,8 @@ interface LP {
     binaries?: string[],
-    generals?: string[]
+    generals?: string[],
+    settings?: {}
```
I am not very familiar with typescript, but we should probably add in here the fields this object can take for now, which should be 3.

rregue commented on glpk.d.ts:
```diff
@@ -26,7 +27,8 @@ interface Result {
         z: number;
-        vars: {[key:string]: number}
+        vars: {[key:string]: number};
+        cons: {[key:string]: {dual: number}}
```
Curious, what other things are you thinking that could be added other than the dual value?

saeid3 replied:
Constraints have several attributes that can be added. Like slack value, basis status, simplex start vector, whether it is lazy constraint, RHS sensitivity information to name a few

rregue commented on src/pre.js:
```diff
@@ -211,24 +214,27 @@ var glpkPromise = new Promise(function (resolve) {
+				
+				// this condition checks if the problem has binary or int columns
+				if (glp_get_num_int(P) || glp_get_num_bin(P)) { 
+					solve_mip(P, settings.msgLev, settings.mipGap, settings.tmLim);
```
what happens if mipGap and tmLim are not set? Same is true for msgLev. Should we set some defaults or is this something that already happens on the C side?

saeid3 replied:
It doesn't happen on C side, I added default value in the glpk side. so if we don't provide settings it'll assign default values.

rregue commented on test/test.js:
```diff
@@ -14,7 +14,12 @@ require('../glpk.js').then(glpkjs => {
 			let prob = new glpk.Problem();
-			let z1, z2 = glpkjs.solve(json, glpkjs.GLP_MSG_ERR).result.z;
+			const settings = {
+        msgLev: glpkjs.GLP_MSG_ERR,
```
the indentation here seems off.

rregue replied:
also, I made a comment around default values and what happens when some of the values are not set. We should decide if we make them optional or not.

rregue reviewed (changes requested):
I left minor comments, mostly around documentation and asking around how default and empty values are being managed.

saeid3 committed:
make setting optional and set default value for it

saeid3 commented:
with https://github.com/TakeScoop/glpk.js/pull/1/commits/0b2585173fe6096ec3d5e112e4d63f7f8c54cb69 I made the `settings` optional and set default value to it. Also I changed the `tm_lim` type to int & `mip_gap` to double as what it is in original C code of [glpk](http://www.chiark.greenend.org.uk/doc/glpk-doc/glpk.pdf).
I updated the typescript accordingly

rregue commented on src/pre.js:
```diff
@@ -90,21 +94,28 @@ var glpkPromise = new Promise(function (resolve) {
 				},
-				'solve': function (lp, msg_lev) {
-
+				'solve': function (lp, jsonSettings) {
```
maybe we can call this options, instead of jsonSettings or `_settings`

rregue commented on src/pre.js:
```diff
@@ -90,21 +94,28 @@ var glpkPromise = new Promise(function (resolve) {
+				'solve': function (lp, jsonSettings) {
+					
+					const settings = _.defaults(jsonSettings, {
+							msgLev: 1,
```
can't we use GLP_MSG_ERR?

saeid3 replied:
No we can't here. it's just defined as a key of API obj and solve itself is another key.

rregue replied:
I see, this unfortunate... so maybe a comment here should be added to know from where the 1 comes from.

rregue commented on test/test.js:
```diff
@@ -14,7 +14,8 @@ require('../glpk.js').then(glpkjs => {
 			let json = JSON.parse(fs.readFileSync(`${__dirname}/data/${d}.json`).toString());
 			let prob = new glpk.Problem();
-			let z1, z2 = glpkjs.solve(json, glpkjs.GLP_MSG_ERR).result.z;
+			const settings = json.settings
```
Should we add a test to test that the settings work as expected? Maybe we can a medium sized problem and set the tmLim to 0 and see if it works.

saeid3 replied:
It's a good idea but since I have removed the building process, it means the changes haven't taken place. So in another PR after we build the changes, we can create new tests.

rregue replied:
You are totally right

rregue reviewed (changes requested):
I just left a couple of thoughts and suggestions. I think however, that we should remove files generated by the built process, as it is unclear if they have been updated on every commit. We can have a separate PR just for the build. 

The other suggestion is around adding more tests to make sure that the settings we are passing in are working as expected.

saeid3 committed:
undo compilation

saeid3 committed:
change jsonSettings to options

saeid3 commented:
I removed the compilations from this PR (we'll have them in separate PR)

rregue reviewed (approved):
Looks good!

Saeid3 commented:
@rregue can you just approve then?
